core/example/tutorial/07_sorting/sorting.cpp

```diff
@@ -0,0 +1,181 @@
+      some algorithms, this data structure may be all that is necessary if the
+      data does not have to be physically reordered.
+
+      Also note here that we are going to reorder the entire AoSoA. There are
+      more advanced variations on sorting which will only operate on a
+      specified range of tuples through a begin and end argument (to be
+      covered in a later tutorial).
+     */
+    auto keys = aosoa.slice<1>();
+    auto sort_data = Cabana::sortByKey( keys );
+
+    /*
+      Now actually permute the AoSoA (i.e. reorder the data).
+    */
+    Cabana::permute( sort_data, aosoa );
+
+    /*
+       Now let's read the data we just sorted. The integer member should
+       appear in ASCENDING ORDER to indicate the sorting has been
+       completed. The rest of the data will also have been sorted as well -
+       all data in each tuple is permuted.
+     */
+    for ( int t = 0; t < num_tuple; ++t )
+    {
+        auto tp = aosoa.getTuple( t );
+
+        // Should now be in DESCENDING ORDER!
+        std::cout << "Tuple " << t
+                  << ", member 0: " << tp.get<0>() << std::endl;
+
+        // Should now be in ASCENDING ORDER!
+        std::cout << "Tuple " << t
+                  << ", member 1: " << tp.get<1>() << std::endl;
+    }
+    std::cout << std::endl;
+
+    /*
+      Now let's demonstrate binning. Binning groups keys in ascending order
+      (if not custom comparator) in a fixed number of bins. The width of the
+      bins (i.e. the largest and smallest values that will qualify as members
+      of the bin) are computed based on a user-assigned number of bins and the
+      minimum and maximum key values.
+
+      Start by creating some new keys. We will manually create an alternating
+      key pattern to demonstrate the effects of binning:
+    */
+    keys(0) = 100;
+    keys(1) = 200;
+    keys(2) = 100;
+    keys(3) = 200;
+    keys(4) = 100;
+
+    /*
+       Now create the binning data. In this case let's create two bins: one
+       for the 100's and one for the 200's.
+
+       Again note that there are variants of binning which can bin values over
+       a subset of the entire AoSoA by using begin and end parameters. We are
+       going to bin the entire AoSoA in this case.
+    */
+    int num_bin = 2;
+    auto bin_data = Cabana::binByKey( keys, num_bin );
+
+    /*
+      Now permute the data. This time the data should be grouped by key
+      value.
+    */
+    Cabana::permute( bin_data, aosoa );
+
+    /*
+       Now let's read the data we just sorted. The integer member should
+       be ordered in binned groups.
+     */
+    for ( int t = 0; t < num_tuple; ++t )
+    {
+        auto tp = aosoa.getTuple( t );
+
+        std::cout << "Tuple " << t
+                  << ", member 0: " << tp.get<0>() << std::endl;
+
+        std::cout << "Tuple " << t
+                  << ", member 1: " << tp.get<1>() << std::endl;
+    }
+    std::cout << std::endl;
+}
+
```

Comment:
Is there a way to get the number of values in each bin?

Reply:
Yes - the sorting/binning returns a `Cabana::BinningData` object. The function `bin_data.binSize( bin_id )` gets the number of values in the bin with the given id while the permutation vector gives the bins themselves and offsets tells you where each bin starts in the permutation vector.

I'll add some to the example explaining how to get the binning data information.